Utils/Elasticsearch/Oracle2JSON.py

```diff
@@ -20,44 +22,141 @@ def connectDEFT_DSN(dsn):
+        step_hours = int(config.get("timestamps", "step_hours"))
+        final_date = config.get("timestamps", "final")
+        if final_date == '':
+            final_date = datetime.now().strftime('%d-%m-%Y %H:%M:%S')
+        queries_cfg = config.items("queries")
+        queries = {}
+        for (qname, file) in queries_cfg:
+            queries[qname] = {'file': file}
+        # changeable data
+        offset_date = config.get("timestamps", "offset")
+        if offset_date == '':
+            offset_date = initial_date
     except IOError:
-        sys.stderr.write('Could not read config file %s' % args.config)
+        sys.stderr.write('Could not read config file %s' % conf)
 
-    global dsn
-    dsn = Config.get("oracle", "dsn")
     conn, cursor = connectDEFT_DSN(dsn)
+    process(conn, offset_date, final_date, step_hours, queries)
+
+
+def plain(conn, queries, offset_date, end_date):
+    tasks = query_executor(conn, queries['tasks']['file'], offset_date, end_date)
+    for task in tasks:
+        task['phys_category'] = get_category(task)
+        # send NDJSON string to STDOUT
+        sys.stdout.write(json.dumps(task) + '\n')
+
+
+def squash(conn, queries, offset_date, end_date):
+    tasks = query_executor(conn, queries['tasks']['file'], offset_date, end_date)
+    datasets = query_executor(conn, queries['datasets']['file'], offset_date, end_date)
+    join_results(tasks, squash_records(datasets))
+
+def squash_records(rec):
+    """
+    a single-pass iterator (for a generator) restructuring list of datasets: 
+    from view: 
+    [
+       {"taskid": 1, "type": "input", "datasetname": "First_in"},
+       {"taskid": 1, "type": "input", "datasetname": "Second_in"},
+       {"taskid": 1, "type": "input", "datasetname": "Third_in"},
+       {"taskid": 1, "type": "output", "datasetname": "First_out"},
+       {"taskid": 1, "type": "output", "datasetname": "Second_out"},
+       {"taskid": 1, "type": "output", "datasetname": "Third_out"},
+       {"taskid": 2, "type": "input", "datasetname": "First_in"},
+       {"taskid": 2, "type": "output", "datasetname": "First_out"},
+       {"taskid": 2, "type": "output", "datasetname": "Second_out"},
+    ...]
+     to 
+    [
+       {"taskid": 1, 
+        "input": ["First_in", "Second_in", "Third_in"], 
+        "output": ["First_out", "Second_out", "Third_out"]
+       },
+       {"taskid": 2, 
+        "input": ["First_in"], 
+        "output": ["First_out","Second_out"]
+       },
+    ...]
+    """
+    grouper = defaultdict(lambda: defaultdict(list))
+    for d in rec:
+        grouper[d['taskid']][d['type']].append(d['datasetname'])
+    return [dict(taskid=k, **v) for k, v in grouper.items()]
+
+
+def join_results(tasks, datasets):
+    d = defaultdict(dict)
+    for l in (tasks, datasets):
+        for elem in l:
+            d[elem['taskid']].update(elem)
+            sys.stdout.write(json.dumps(str(d[elem['taskid']])) + '\n')
```

Comment:
The following stage (019) fails to decode output:
`str(d[elem['taskid']])` returns a string, not a `dict` -- and when we apply `json.dumps()` to a string, we have an original string in double quotes instead of JSON document.